mozilla_django_oidc_db/jwt.py

```diff
@@ -0,0 +1,53 @@
+"""
+Support for user info JWT verification and decoding.
+
+The bulk of the implementation is taken from mozilla-django-oidc where the access token
+is processed, but adapted for non-hardcoded/configured parameters.
+
+In the case of Keycloak for example, the token signing algorithm is configured on the
+server and can change on a whim.
+"""
+
+import json
+from typing import Any
+
+from django.core.exceptions import SuspiciousOperation
+from django.utils.encoding import smart_bytes
+
+from josepy.jwk import JWK
+from josepy.jws import JWS
+
+
+def verify_and_decode_token(token: bytes, key) -> dict[str, Any]:
+    """
+    Verify that the token was not tampered with and if okay, return the payload.
+
+    This is mostly taken from
+    :meth:`mozilla_django_oidc.auth.OIDCAuthenticationBackend._verify_jws`.
+    """
+
+    jws = JWS.from_compact(token)
+
+    # validate the signing algorithm
+    if (alg := jws.signature.combined.alg) is None:
+        raise SuspiciousOperation("No alg value found in header")
+
+    # one of the most common implementation weaknesses -> attacker can supply 'none'
+    # algorithm
+    if alg.name == "none":
+        raise SuspiciousOperation("'none' for alg value is not allowed")
+
+    # process key parameter which was/may have been loaded from keys endpoint. I'm u
+    # Unsure what the type of this parameter can be :/
```

Comment:
The key seems to be retrieved from the JWKS endpoint, so it should be in the JWKSet format (https://datatracker.ietf.org/doc/html/rfc7517#section-5). as `retrieve_matching_jwk` iterated over the keys to get the correct one, `key` should be in the JWK format (i.e. a dict).

However, I see the upstream implementation also takes into accounts string :shrug:

Reply:
Well the JWKS endpoint is supposed to be public so if people want to make their private HS* keys available they probably can even if it's a really bad idea :):)

Reply:
HMAC relies on out-of-band secret exchange so it's probably a valid approach but unsupported for us here. I don't expect this to be a mechanism to ever be used by a big vendor or solution like Keycloak.